### PR TITLE
[ECE] Add button radius support

### PR DESCRIPTION
**Pull request by reykjalin**
## Changes proposed in this Pull Request

* Adds support for changing the border radius on the ECE and WooPay express checkout buttons.

**Questions for design:**

- Is the slider (i.e. arbitrary pixel value) preferable, or something like a "none/small/medium/large" radio button selection?
    - The "none/small/medium/large" can also be implemented on a slider.
- Should we be following the outcome from https://github.com/woocommerce/woocommerce/discussions/48055 for this instead of applying this from the WooPayments settings?
    - No, we need the backwards compatibility.

### Demo

https://github.com/Automattic/woocommerce-payments/assets/13835680/1879f085-d2b8-4361-b1cc-404d4e514abb

## Testing instructions

* Test different border radii with ECE enabled and ensure they are applied correctly to all ECE and WooPay buttons on Shortcode and Blocks pages.
* Disable ECE and ensure all buttons are rendered with the default radius.

### Example tests to run

> [!IMPORTANT]
> Please run various combinations of these tests. The following are examples you can build upon.

#### Test the default value

1. Enable ECE; `npm run wp option set _wcpay_feature_stripe_ece 1`.
2. Check that the radius is `4px` on all pages &mdash; Product, Cart (Blocks + Shortcode), Checkout (Blocks + Shortcode) &mdash; and to all buttons &mdash; WooPay, Google Pay, and Apple Pay.
    - **Note:** Link is not implemented using ECE yet, and as such does not support setting the border radius.

#### Test border radii

1. Enable ECE; `npm run wp option set _wcpay_feature_stripe_ece 1`.
3. Set the border radius to the maximum value.
4. Check that radius is applied on all pages &mdash; Product, Cart (Blocks + Shortcode), Checkout (Blocks + Shortcode) &mdash; and to all buttons &mdash; WooPay, Google Pay, and Apple Pay.
    - **Note:** Link is not implemented using ECE yet, and as such does not support setting the border radius.
5. Set the border radius to 0 and repeat step (3).
6. Set the border radius to some value in between 0 and the max and repeat step (3).

#### Test PRBs

1. Set the border radius to the maximum value.
2. Disable ECE; `npm run wp option set _wcpay_feature_stripe_ece 1`.
3. Check that the max radius is not applied on any pages &mdash; Product, Cart (Blocks + Shortcode), Checkout (Blocks + Shortcode) &mdash; and to all buttons &mdash; WooPay, Google Pay, and Apple Pay.
    - **Note:** Link is not implemented using ECE yet, and as such does not support setting the border radius.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by reykjalin** [comment]
@elizaan36 @nikkivias would love to get your opinion on the approach here, especially for the questions in the PR description.

File: client/data/settings/selectors.js [hunk 1]
```diff
@@ -184,6 +184,10 @@ export const getPaymentRequestButtonTheme = ( state ) => {
 	return getSettings( state ).payment_request_button_theme || '';
 };
 
+export const getPaymentRequestButtonBorderRadius = ( state ) => {
+	return getSettings( state )?.payment_request_button_border_radius ?? 10;
```
Comment (reykjalin) [inline client/data/settings/selectors.js]: When ECE is disabled that implies that the Payment Request Buttons are enabled instead. Those do not support changing the border radius, so we need a fallback value when ECE is disabled.

A default radius of 10px was something I just chose arbitrarily while working on the PR. We'll probably want some other default radius for when ECE is disabled, likely to match what the PRBs use.

File: client/express-checkout/utils/index.ts [hunk 2]
```diff
@@ -108,3 +111,70 @@ export const getErrorMessageFromNotice = ( notice: string ) => {
...
+
+/**
+ * Returns the appearance settings for the Express Checkout buttons.
+ * Currently only configures border radius for the buttons.
+ */
+export const getExpressCheckoutButtonAppearance = () => {
+	const buttonSettings = getExpressCheckoutData( 'button' );
+
+	return {
+		variables: { borderRadius: `${ buttonSettings?.radius ?? 10 }px` },
```
Comment (reykjalin) [inline client/express-checkout/utils/index.ts]: When ECE is disabled that implies that the Payment Request Buttons are enabled instead. Those do not support changing the border radius, so we need a fallback value when ECE is disabled.

A default radius of 10px was something I just chose arbitrarily while working on the PR. We'll probably want some other default radius for when ECE is disabled, likely to match what the PRBs use.

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9010` or branch name `add/ece-button-radius-support` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/ece-button-radius-support) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: c35d2d0ecd4d198280109a9e85d68072f8b7864c
- Build time: 2024-07-05 05:54:41 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/class-wc-payments-woopay-button-handler.php [hunk 3]
```diff
@@ -333,6 +333,7 @@ public function display_woopay_button_html() {
 		}
 
 		$settings = $this->get_button_settings();
+		$radius   = WC_Payments_Features::is_stripe_ece_enabled() ? $settings['radius'] : 10;
```
Comment (reykjalin) [inline includes/class-wc-payments-woopay-button-handler.php]: When ECE is disabled that implies that the Payment Request Buttons are enabled instead. Those do not support changing the border radius, so we need a fallback value when ECE is disabled.

A default radius of 10px was something I just chose arbitrarily while working on the PR. We'll probably want some other default radius for when ECE is disabled, likely to match what the PRBs use.

**Review by reykjalin — COMMENTED** [review]
Noting where we're falling back to a `10px` default border radius and why we'll likely want to change that.

File: includes/admin/class-wc-rest-payments-settings-controller.php [hunk 4]
```diff
@@ -470,60 +475,61 @@ public function get_settings(): WP_REST_Response {
...
+				'account_branding_secondary_color'     => $this->wcpay_gateway->get_option( 'account_branding_secondary_color' ),
+				'account_domestic_currency'            => $this->wcpay_gateway->get_option( 'account_domestic_currency' ),
+				'is_payment_request_enabled'           => 'yes' === $this->wcpay_gateway->get_option( 'payment_request' ),
+				'is_debug_log_enabled'                 => 'yes' === $this->wcpay_gateway->get_option( 'enable_logging' ),
+				'payment_request_enabled_locations'    => $this->wcpay_gateway->get_option( 'payment_request_button_locations' ),
+				'payment_request_button_size'          => $this->wcpay_gateway->get_option( 'payment_request_button_size' ),
+				'payment_request_button_type'          => $this->wcpay_gateway->get_option( 'payment_request_button_type' ),
+				'payment_request_button_theme'         => $this->wcpay_gateway->get_option( 'payment_request_button_theme' ),
+				'payment_request_button_border_radius' => WC_Payments_Features::is_stripe_ece_enabled() ? $this->wcpay_gateway->get_option( 'payment_request_button_border_radius' ) : 10,
```
Comment (reykjalin) [inline includes/admin/class-wc-rest-payments-settings-controller.php]: When ECE is disabled that implies that the Payment Request Buttons are enabled instead. Those do not support changing the border radius, so we need a fallback value when ECE is disabled.

A default radius of 10px was something I just chose arbitrarily while working on the PR. We'll probably want some other default radius for when ECE is disabled, likely to match what the PRBs use.

**Comment by reykjalin** [comment]
@bborman22 Sort of took this on as part of #8951 and spent a day on it. Added an estimate and added it to the current sprint 👍 . If you think this should be considered part of #8951 instead, feel free to remove the points estimate.

**Comment by nikivias** [comment]
Hi @reykjalin thanks for the solution you have come up with. I would only suggest small edits to align with the blocks editor styling, like so:

CleanShot 2024-07-01 at 16 24 31@2x

[See it on Figma](https://www.figma.com/design/LSwxXIhP15A1YvZWnkgihH/Payments-Settings-Explorations-2024?node-id=2419-26973&t=4Ab2dVQXP8SXNDoX-1)

#48055 is a cool solution but it is only limited to the checkout blocks. So it may still be important to have this function. cc: @pierorocca

**Comment by pierorocca** [comment]
Definitely lets see if we can fit in this work for Shortcode sites. Blocks I think we need to coordinate with the platform team and the work they're doing on refactoring express checkout settings?

Acceptance criteria: Default value should reflect the current value of 4px so as not to alter existing sites when enabled.

@nikkivias should 0 or 4px be the smallest size allowable in your opinion?

File: includes/admin/class-wc-rest-payments-settings-controller.php (hunk 4 above)
Comment (reykjalin) [reply includes/admin/class-wc-rest-payments-settings-controller.php]: The default radius [should be 4px](https://github.com/Automattic/woocommerce-payments/pull/9010#issuecomment-2200514880).

**Comment by nikivias** [comment]
> @nikkivias should 0 or 4px be the smallest size allowable in your opinion?

@pierorocca 0px, in case that happens to be how they style their buttons.

File: client/data/settings/selectors.js (hunk 1 above)
Comment (reykjalin) [reply client/data/settings/selectors.js]: Fixed in https://github.com/Automattic/woocommerce-payments/pull/9010/commits/b76d492b54b2137d1193677bfc225d9e5dd0658e

File: client/express-checkout/utils/index.ts (hunk 2 above)
Comment (reykjalin) [reply client/express-checkout/utils/index.ts]: Fixed in https://github.com/Automattic/woocommerce-payments/pull/9010/commits/b76d492b54b2137d1193677bfc225d9e5dd0658e

File: includes/class-wc-payments-woopay-button-handler.php (hunk 3 above)
Comment (reykjalin) [reply includes/class-wc-payments-woopay-button-handler.php]: Fixed in https://github.com/Automattic/woocommerce-payments/pull/9010/commits/b76d492b54b2137d1193677bfc225d9e5dd0658e

File: includes/admin/class-wc-rest-payments-settings-controller.php (hunk 4 above)
Comment (reykjalin) [reply includes/admin/class-wc-rest-payments-settings-controller.php]: Fixed in https://github.com/Automattic/woocommerce-payments/pull/9010/commits/b76d492b54b2137d1193677bfc225d9e5dd0658e

**Comment by reykjalin** [comment]
@nikkivias I've updated the UI to be a closer match to the block editor, could you take another look? The video in the description has been updated with the latest implementation.

The `px` label in the UI is hard-coded. I could make it so the merchants can select a unit via drop-down like in the block editor, it'd just take a bit of extra work to implement. I figured I'd get a review on the bare-bones version and see if we want to invest further in the configuration and add support for different units.

**Review by nikivias — APPROVED** [review]
Looks perfect! Thanks

**Review by bborman22 — APPROVED** [review]
This looked and worked great for me. Thanks for addressing that early comment about having the default radius stored in one place, that will make any updates a lot easier! I tested this in a variety of scenarios with different button combos and browsers and also confirmed no radius is applied when ECE is disabled. 

One thing I noticed was that the preview doesn't adjust the radius on the Google Pay button, but I believe this will be addressable in #9028 when we use ECE in the settings preview, so no concern there.

LGTM!

**Comment by pierorocca** [comment]
Looks great. Nice work all.

**Comment by reykjalin** [comment]
> One thing I noticed was that the preview doesn't adjust the radius on the Google Pay button, but I believe this will be addressable in https://github.com/Automattic/woocommerce-payments/pull/9028 when we use ECE in the settings preview, so no concern there.

Yes, that's because in this PR the previews are still using the PRBs, so the border radius update only work when previewing the WooPay button :)

**Pull request merged**